docs/source/tools/mock_trafficops.rst

```diff
@@ -0,0 +1,99 @@
+..
+..
+.. Licensed under the Apache License, Version 2.0 (the "License");
+.. you may not use this file except in compliance with the License.
+.. You may obtain a copy of the License at
+..
+..     http://www.apache.org/licenses/LICENSE-2.0
+..
+.. Unless required by applicable law or agreed to in writing, software
+.. distributed under the License is distributed on an "AS IS" BASIS,
+.. WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+.. See the License for the specific language governing permissions and
+.. limitations under the License.
+..
+
+.. _mock-to:
+
+***************************
+The Mock Traffic Ops Server
+***************************
+.. versionadded:: 4.0.0
+
+.. warning:: At the time of this writing, the Traffic Ops Mock Server is still under active development. As such, it should not be considered stable and/or built into critical procedures until it has reached maturity.
+
+Developers who write clients that consume the :ref:`to-api` may want to write tests for said clients. Currently, this is most typically done by defining a static "data set" for Traffic Ops and loading it into the Traffic Ops Database. This method has two flaws:
+
+* It still requires an authentication handshake at startup to test any given route (though in most cases testing that would also be desired).
+* The :ref:`to-api` produces non-deterministic output, especially for configuration files\ [#configfiles_issue]_. Because of this, no two requests to the same endpoint with the same data in the Traffic Ops Database can be guaranteed to produce the same output.
+
+If these factors are important to the tests for a client of the :ref:`to-api`, it is suggested that they instead utilize the Traffic Ops Mock server. The source code for this binary can be found in :file:`traffic_ops/client_tests`.
+
+Behavior
+========
+The Traffic Ops mock server **SHALL**
+
+* Serve all of the API routes supported by the :ref:`to-api`\ [#implemented_routes_disclaimer]_
+* Serve static responses that never change in structure or content
```

Comment:
One of the most valuable properties of a mock is the ability to easily simulate failures. Could we consider that a potential use for the mock TO? Specifically, can we have some routes that are hardcoded to specific types of failures?

Reply:
Could you describe a planned failure? I'm not sure what you mean by that.

Reply:
Something like a route which would return a 404 or malformed response (not JSON). I'm not totally sure how(unit test, integration test, etc...) you are planning to use this mock server, so it might not make sense. Here's an example of some negative test cases we could run against the ORT script:

* How does ORT handle a TO timeout?
* How does ORT handle getting a 404 from a route it thinks should exist?
* What happens if ORT gets not JSON (or not the expected format) somewhere (we've seen garbage end up in some of the ATS config files....) 
* What if ORT gets a remap activating a plugin that isn't present (something like this would be best handled in a unit test for the remap.config or plugin.config functions. I don't know if I'd put something this specific into an integration test).

Reply:
I had designed this with only positive tests in mind. Negative tests are, of course, valuable, but I'm not sure how to intertwine the two while the server is separate from the test suite. How does the test suite know which routes are supposed to 404? If we build tests into the mock server, it destroys the ability to use it generically, because now it's only for testing something specific.